test-poc/framework/README.md

```diff
@@ -0,0 +1,338 @@
+}
+```
+
+When the lifecycle is set to Method the realm is automatically destroyed and re-created for each test method, as seen in 
+the above example where one test method adds a user to the realm, but the user is not present in the next test.
+
+The general recommendation is to use the Class lifecycle for realms, clients, and users. Making sure that individual test
+methods leave the resource in a way that can be re-used. Realms for example with global lifecycle can be harder to
+maintain as individual test classes can break other tests, but at the same time using global resources can be useful 
+as it will be more performant.
+
+## Configuring resources
+
+Resources are configured by declaring the required configuration through a Java class. This Java class can be an inner-class
+if it's only used for a single test class, or can be a proper class when multiple tests share the same configuration.
+
+For example to create a realm with a specific configuration:
+
+```java
+@InjectRealm(config = MyRealmConfig.class)
+ManagedRealm realm;
+
+static class MyRealmConfig implements RealmConfig {
+
+    @Override
+    public RealmRepresentation getRepresentation() {
+        return builder()
+                .name("myrealm")
+                .groups("group-a", "group-b")
+                .build();
+    }
+}
+```
+
+The framework will automatically re-create global resources if they don't match the required configuration. For example:
+
+```java
+@KeycloakIntegrationTest
+public class Test1 {
+
+    @InjectRealm(lifecycle = LifeCycle.GLOBAL, config = MyRealmConfig.class)
+    ManagedRealm realm;
+
+}
+
+@KeycloakIntegrationTest
+public class Test2 {
+
+    @InjectRealm(lifecycle = LifeCycle.GLOBAL, config = MyOtherRealm.class)
+    ManagedRealm realm;
+
+}
+```
+
+In this example the realm from `Test1` would be destroyed and a new realm created for `Test2` since different 
+configuration is requested.
+
+## Multiple instances
+
+By default, a resource does not have a reference, and child-resources are created within the resource without a reference.
+For example in the following example `userA` will be created within `realmA`:
+
+```java
+@InjectRealm
+ManagedRealm realmA;
+
+@InjectUser
+ManagedUser userA;
+```
+
+If you need for instance multiple realms within a test you need to set a reference on it, and use this reference for
+child resources:
+
+```java
+@InjectRealm
+ManagedRealm realmA;
+
+@InjectUser
+ManagedUser userA;
+
+@InjectRealm(ref = "realmB")
+ManagedRealm realmB;
+
+@InjectUser(realmRef = "realmB")
+ManagedUser userB;
+```
+
+Resources using a ref can of course also have a Global lifecycle, and will also automatically be dropped if the requested configuration differs. This will be done on a per-ref basis. For example if two tests use the same ref for two global realms, but only one has different configuration, then only the realm with the ref with an incompatible config will be re-created.
+
```

Comment:
This sentence is a bit confusing. I had to read it a few times until I understood it. Perhaps another example could help, wdyt? Something like this:
```suggestion
```java
@KeycloakIntegrationTest
public class Test1 {

    @InjectRealm(lifecycle = LifeCycle.GLOBAL, ref = "myRealm1", config = MyRealmConfig.class)
    ManagedRealm realm1;

    @InjectRealm(lifecycle = LifeCycle.GLOBAL, ref="myRealm2", config = MyOtherRealm.class)
    ManagedRealm realm2;
}

@KeycloakIntegrationTest
public class Test2 {

    @InjectRealm(lifecycle = LifeCycle.GLOBAL, ref = "myRealm1", config = MyRealmConfig.class)
    ManagedRealm realm1;
    
    // "myRealm1" from Test1 is reused

    @InjectRealm(lifecycle = LifeCycle.GLOBAL, ref="myRealm2", config = TotallyDifferentRealm.class)
    ManagedRealm realm2;

     // "myRealm2" from Test1 is dropped and re-created
}
```
```

Reply:
Re-wrote this and included your example, but a little re-worked to make it a bit more consistent with the other examples.